src/2025h1/unsafe-fields.md

```diff
@@ -0,0 +1,118 @@
+        self.len += 2;
+    }
+}
+```
+
+Rust cannot enforce that modifications of `len` require `unsafe`, because the language does not provide the programmer a way of communicating to the compiler that `len` carries safety invariants.
+
+[*The Scope of Unsafe*]: https://www.ralfj.de/blog/2016/01/09/the-scope-of-unsafe.html
+
+### The "shiny future" we are working towards
+
+Rust programmers will use the `unsafe` keyword to denote fields that carry library safety invariants; e.g.:
+
+```rust
+struct Vec<T> {
+    // SAFETY: The elements `data[i]` for
+    // `i < len` are in a valid state.
+    unsafe data: Box<[MaybeUninit<T>]>,
+    unsafe len: usize,
+}
+```
+
+Rust will require that usages of `unsafe` fields which could violate their safety invariants must *only* occur within `unsafe` contexts.
+
+### The next 6 months
+
+In the next six months, we will iterate on the design and implementation of unsafe fields. An RFC for `unsafe` fields will be accepted, and a candidate implementation will — at the very least — be ready to enter the stabilization process.
+
+## Design axioms
+
+The design of `unsafe` fields is guided by three axioms:
+
+1. **Unsafe Fields Denote Safety Invariants**
+   A field *should* be marked `unsafe` if it carries arbitrary library safety invariants with respect to its enclosing type.
+2. **Unsafe Usage is Always Unsafe**
+   Uses of `unsafe` fields which could violate their invariants *must* occur in the scope of an `unsafe` block.
+3. **Safe Usage is Usually Safe**
+   Uses of `unsafe` fields which cannot violate their invariants *should not* require an unsafe block.
+
+## Ownership and team asks
+
+**Owner:** @jswrenn
+
+| Subgoal                                        | Owner(s) or team(s)  | Notes                          |
+| ---------------------------------------------- | -------------------- | ---------                      |
+| Discussion and moral support                   | ![Team][] [lang]     | [Zulip]                        |
+| Stabilize Unsafe Fields                        |                      |                                |
+| ↳ Author RFC                                   | @jhpratt             | [RFC3458], [Living Design Doc] |
+| ↳ Implementation                               | @veluca93            |                                |
+| ↳ Standard reviews                             | ![Team][] [compiler] |                                |
+| ↳ Design meeting                               | ![Team][] [lang]     |                                |
+| ↳ RFC decision                                 | ![Team][] [lang]     |                                |
+| ↳ Secondary RFC review                         | ![Team][] [lang]     |                                |
+| ↳ Author stabilization report                  | @jswrenn             |                                |
+| ↳ Stabilization decision                       | ![Team][] [lang]     |                                |
```

Comment:
```suggestion
```

Based on your notes it doesn't sound like we expect to stabilize in the next 6 months